cli/registry.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+)
+
+// defaultRegistriesGlobalConfig is a global registries file that will be used in all the servers/workers
+const defaultRegistriesGlobalConfig = "$HOME/.k3d/registries.yaml"
+
+const defaultRegistryContainerName = "registry"
+
+const defaultRegistryImage = "registry:2"
+
+const defaultFullRegistriesPath = "/etc/rancher/k3s/registries.yaml"
+
+// NOTE: structs copied from https://github.com/rancher/k3s/blob/master/pkg/agent/templates/registry.go
+//       for avoiding a dependencies nightmare...
+
+// Registry is registry settings configured
+type Registry struct {
+	// Mirrors are namespace to mirror mapping for all namespaces.
+	Mirrors map[string]Mirror `toml:"mirrors" yaml:"mirrors"`
+
+	// Configs are configs for each registry.
+	// The key is the FDQN or IP of the registry.
+	Configs map[string]interface{} `toml:"configs" yaml:"configs"`
+
+	// Auths are registry endpoint to auth config mapping. The registry endpoint must
+	// be a valid url with host specified.
+	// DEPRECATED: Use Configs instead. Remove in containerd 1.4.
+	Auths map[string]interface{} `toml:"auths" yaml:"auths"`
+}
+
+// Mirror contains the config related to the registry mirror
+type Mirror struct {
+	// Endpoints are endpoints for a namespace. CRI plugin will try the endpoints
+	// one by one until a working one is found. The endpoint must be a valid url
+	// with host specified.
+	// The scheme, host and path from the endpoint URL will be used.
+	Endpoints []string `toml:"endpoint" yaml:"endpoint"`
+}
+
+// writeRegistriesConfigInContainer creates a valid registries configuration file in a container
+func writeRegistriesConfigInContainer(spec ClusterSpec, ID string) error {
+	globalRegistriesConfig := os.ExpandEnv(defaultRegistriesGlobalConfig)
+	registryFullName := fmt.Sprintf("%s:%d", spec.RegistryName, spec.RegistryPort)
+
+	privRegistries := &Registry{}
+
+	// if ${HOME}/.k3d/registries.yaml exists, load it
+	privRegistryFile, err := ioutil.ReadFile(globalRegistriesConfig)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+	} else {
+		if err := yaml.Unmarshal([]byte(privRegistryFile), &privRegistries); err != nil {
+			return err
+		}
+	}
+
+	if len(privRegistries.Mirrors) == 0 {
+		privRegistries.Mirrors = map[string]Mirror{}
+	}
+
+	// the add the private registry
+	privRegistries.Mirrors[registryFullName] = Mirror{
+		Endpoints: []string{fmt.Sprintf("http://%s", registryFullName)},
+	}
+
+	d, err := yaml.Marshal(&privRegistries)
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	docker, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return fmt.Errorf("Couldn't create docker client\n%+w", err)
+	}
+
+	return docker.CopyToContainer(ctx, ID, defaultFullRegistriesPath, bytes.NewReader(d), types.CopyToContainerOptions{
```

Comment:
An addition here from the Documentation of `CopyToContainer`: 
> Note that `content` must be a Reader for a TAR archive

Reply:
Yep, you are right, but I think that code already does that (as I have just taken it from the docker client code). Anyway I think it can be simplified...